streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -594,131 +664,158 @@ public GroupAssignment assign(final Cluster metadata, final GroupSubscription gr
                         minSupportedMetadataVersion,
                         assignedActiveList,
-                        standbyTasks,
+                        standbyTaskMap,
                         partitionsByHostState,
-                        0)
-                        .encode()
-                ));
-            }
-        }
-
-        // add empty assignment for "future version" clients (ie, empty version probing response)
-        for (final String consumerId : futureConsumers) {
-            assignment.put(consumerId, new Assignment(
-                Collections.emptyList(),
-                new AssignmentInfo(minUserMetadataVersion, minSupportedMetadataVersion).encode()
-            ));
+                        AssignorError.NONE.code()
+                    ).encode()
+                )
+            );
         }
-
-        return assignment;
     }
 
-    private static void buildAssignedActiveTaskAndPartitionsList(final List<TaskId> activeTasks,
-                                                                 final List<TopicPartition> activePartitionsList,
-                                                                 final List<TaskId> assignedActiveList,
-                                                                 final Map<TaskId, Set<TopicPartition>> partitionsForTask) {
+    private void buildAssignedActiveTaskAndPartitionsList(final String consumer,
+                                                          final ClientState clientState,
+                                                          final List<TaskId> activeTasksForConsumer,
+                                                          final Map<TaskId, Set<TopicPartition>> partitionsForTask,
+                                                          final Set<TopicPartition> allOwnedPartitions,
+                                                          final List<TopicPartition> activePartitionsList,
+                                                          final List<TaskId> assignedActiveList) {
         final List<AssignedPartition> assignedPartitions = new ArrayList<>();
 
         // Build up list of all assigned partition-task pairs
-        for (final TaskId taskId : activeTasks) {
+        for (final TaskId taskId : activeTasksForConsumer) {
             for (final TopicPartition partition : partitionsForTask.get(taskId)) {
-                assignedPartitions.add(new AssignedPartition(taskId, partition));
+                final String oldOwner = clientState.ownedPartitions().get(partition);
+                final boolean newPartitionForConsumer = oldOwner == null || !oldOwner.equals(consumer);
+
+                // If the partition is new to this consumer but is still owned by another, remove from the assignment
+                // until it has been revoked and can safely be reassigned according the COOPERATIVE protocol
+                if (newPartitionForConsumer && allOwnedPartitions.contains(partition)) {
```

Comment:
Is it possible that a task maps to multiple partitions, while only some of them have old owners, would that cause us to encode a partial task here?

Reply:
Ah, I see what you mean. We should collect the assignedPartitions for a single task and only add them to the `assignedPartitions` list at the end, if all can be safely assigned